backend/shard/src/connection.rs

```
@@ -0,0 +1,195 @@
+                            log::error!(
+                                "Error sending message to websocker server (will reconnect): {}",
+                                e
+                            );
+                            break;
+                        }
+                    }
+                }
+                Err(e) => {
+                    // Issue connecting? Wait and try again on the next loop iteration.
+                    log::error!(
+                        "Error connecting to websocker server (will reconnect): {}",
+                        e
+                    );
+                }
+            };
+
+            // Tell the aggregator that we're disconnected so that, if we like, we can discard
+            // messages without doing any futher processing on them.
+            if connected {
+                connected = false;
+                let _ = tx_to_external.send(Message::Disconnected).await;
+            }
+
+            // Wait a little before trying to reconnect.
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        }
+    });
+
+    tx_to_connection_proxy
+}
+
+/// This spawns a connection to a websocket server, serializing/deserialziing
+/// from bincode as messages are sent or received.
+async fn create_ws_connection_no_retry<In, Out, S, E>(
+    mut tx_to_external: S,
+    telemetry_uri: http::Uri,
+) -> anyhow::Result<mpsc::Sender<In>>
+where
+    S: Sink<Message<Out>, Error = E> + Unpin + Send + 'static,
+    E: std::fmt::Debug + std::fmt::Display,
+    In: serde::Serialize + Send + 'static,
+    Out: serde::de::DeserializeOwned + Send + 'static,
+{
+    use bincode::Options;
+    use soketto::handshake::{Client, ServerResponse};
+
+    let host = telemetry_uri.host().unwrap_or("127.0.0.1");
+    let port = telemetry_uri.port_u16().unwrap_or(8000);
+    let path = telemetry_uri.path();
+
+    let socket = TcpStream::connect((host, port)).await?;
+    socket.set_nodelay(true).unwrap();
```

return err?

---

Good point!

---

(I went with `expect` in the end as I'm not expecting it to reasonably fail (and if it does it'll fail fast at least))